### PR TITLE
Adding webhook to prom metrics

### PR DESCRIPTION
**Pull request by bfoley13**
# Description

This changes adds mutation/validation webhooks into our current Prometheus metrics.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration. Is it a breaking change which will impact consuming tool(s)?

- [x] Unit Tests

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: pkg/webhook/nginxingress.go [hunk 1]
```diff
@@ -27,9 +30,15 @@ const (
 	mutationPath   = "/mutate-nginx-ingress-controller"
 )
 
+var (
+	nginxResourceValidationName = controllername.New("nginxIngressResourceValidator")
```
Comment (OliverMKing) [inline pkg/webhook/nginxingress.go:34]: splitting up each word of the controller name is probably best

File: pkg/webhook/nginxingress.go [hunk 2]
```diff
@@ -177,12 +193,13 @@ func (n *nginxIngressResourceValidator) Handle(ctx context.Context, req admissio
 
 	lgr.Info("decoding NginxIngressController resource")
 	var nginxIngressController approutingv1alpha1.NginxIngressController
-	if err := n.decoder.Decode(req, &nginxIngressController); err != nil {
+	if err = n.decoder.Decode(req, &nginxIngressController); err != nil {
```
Comment (OliverMKing) [inline pkg/webhook/nginxingress.go:196]: not related to this pr but could you add a log statement in this block? noticed I missed one

File: pkg/webhook/nginxingress.go [hunk 3]
```diff
@@ -158,10 +168,16 @@ var sarAuthenticate = func(ctx context.Context, lgr logr.Logger, cl client.Clien
...
+	defer func() {
+		metrics.HandleWebhookHandlerMetrics(nginxResourceValidationName, err)
+	}()
+
+	lgr := logr.FromContextOrDiscard(ctx).WithValues("resourceName", req.Name, "namespace", req.Namespace, "operation", req.Operation).WithName(nginxResourceValidationName.LoggerName())
 
 	// ensure user has permissions required
-	cantPerform, err := n.authenticate(ctx, lgr, n.client, req)
+	var cantPerform string
+	cantPerform, err = n.authenticate(ctx, lgr, n.client, req)
```
Comment (OliverMKing) [inline pkg/webhook/nginxingress.go:180]: need to set an error if cantPerform != "" if we are incrementing error count on the Denied response.

Alternatively we can add an additional variable to the `metrics.HandleWebhookHandlerMetrics` fn to take in the admission.Response and increment error if it's denied.

Comment (OliverMKing) [reply pkg/webhook/nginxingress.go:180]: I'd probably recommend the second approach.

Comment (OliverMKing) [reply pkg/webhook/nginxingress.go:180]: I feel like that's a lot easier conceptually than setting errors

File: pkg/webhook/nginxingress.go [hunk 4]
```diff
@@ -238,14 +263,15 @@ func (n nginxIngressResourceMutator) Handle(ctx context.Context, request admissi
 
 	lgr.Info("decoding NginxIngressController resource")
 	var nginxIngressController approutingv1alpha1.NginxIngressController
-	if err := n.decoder.Decode(request, &nginxIngressController); err != nil {
+	if err = n.decoder.Decode(request, &nginxIngressController); err != nil {
 		return admission.Errored(http.StatusBadRequest, fmt.Errorf("decoding NginxIngressController: %w", err))
```
Comment (OliverMKing) [inline pkg/webhook/nginxingress.go:267]: add logging statement here because I didn't

File: pkg/webhook/nginxingress.go [hunk 5]
```diff
@@ -238,14 +263,15 @@ func (n nginxIngressResourceMutator) Handle(ctx context.Context, request admissi
...
 	}
 
 	lgr.Info("defaulting NginxIngressController resource")
 	nginxIngressController.Default()
 
-	marshalled, err := json.Marshal(nginxIngressController)
+	var marshalled []byte
+	marshalled, err = json.Marshal(nginxIngressController)
 	if err != nil {
 		return admission.Errored(http.StatusInternalServerError, fmt.Errorf("marshalling NginxIngressController: %w", err))
```
Comment (OliverMKing) [inline pkg/webhook/nginxingress.go:276]: logging statement in this block too because I didn't

**Review by OliverMKing — COMMENTED** [review]
comment

File: pkg/webhook/nginxingress.go (hunk 3 above)
Comment (bfoley13) [reply pkg/webhook/nginxingress.go:179]: swapped to using admission response as input to metric handler

File: pkg/webhook/nginxingress.go (hunk 2 above)
Comment (bfoley13) [reply pkg/webhook/nginxingress.go:195]: added

File: pkg/webhook/nginxingress.go (hunk 4 above)
Comment (bfoley13) [reply pkg/webhook/nginxingress.go:267]: added

File: pkg/webhook/nginxingress.go (hunk 5 above)
Comment (bfoley13) [reply pkg/webhook/nginxingress.go:277]: added

File: pkg/webhook/nginxingress.go (hunk 1 above)
Comment (bfoley13) [reply pkg/webhook/nginxingress.go]: I have split up the words in each definition

**Review by OliverMKing — APPROVED** [review]

**Pull request merged**